[4.1] Raven db 10538

aviv86 opened the pull request.

aviv86 committed:
RavenDB-10538 : range facet should use parameters
- changed approach: do the parsing of RangeFacet<T>.Ranges when doing the conversion, but delay the conversion to the last possible point (inside FacetToken.Create<T>)
- it covers all cases except creating RangeFacet<T> inside a List<RangeFacet> due to implicit conversion that exists between those two types.

aviv86 committed:
RavenDB-10538 : tests

ayende commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -269,16 +265,19 @@ private static string GetStringRepresentation(string fieldName, ExpressionType l
+
             switch (value.GetType().FullName)
             {
                 //The nullable stuff here it a bit weird, but it helps with trying to cast Value types
```
Need to handle the nullable stuff here as well.
Also, why are we doing string comparisons for the types?

ayende replied:
Let's handle this everything as a parameters, not just dates & strings

ppekrol replied:
String comparision is a left-over from 3.x. I did not notice this when I was implementing Facets in 4.0

ayende commented on src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs:
```diff
@@ -213,6 +213,17 @@ private static ParsedRange ParseRange(QueryExpression expression, FacetQuery que
                     RangeText = expression.GetText()
                 };
+
+                if (bee.Max.Value == ValueTokenType.Parameter)
```
Don't we have a better way of handling this than doing text replace?

ayende commented on src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs:
```diff
@@ -236,6 +247,12 @@ private static ParsedRange ParseRange(QueryExpression expression, FacetQuery que
 
+                        if (r.Value == ValueTokenType.Parameter)
+                        {
+                            range.RangeText = range.RangeText.Replace(r.Token, fieldValue.Value);
```
Same comment

ayende commented on test/SlowTests/Tests/Faceted/FacetsWithParameters.cs:
```diff
@@ -0,0 +1,423 @@
+                    {
+                        //this facet will not use query-parameters when excecuting a query, due to the
+                        //implicit conversion that exists between RangeFacet<Foo> and RangeFacet
+
```
Why does this matter?

ppekrol replied:
This is a case that we cannot support without doing breaking changes. The `RangeFacet<T>` contains an implicit conversion to `RangeFacet`. We should leave it like that.

aviv86 committed:
RavenDB-10538 :
- support using parameters for RangeFacet<T> inside List<RangeFacet>
- send numbers as parameters as well
- RangeFacet<T>.GetStringValue() : avoid doing string comparisons for the types
- NumberUtil.cs - add NumberToString(float number) overload

aviv86 committed:
RavenDB-10538 :
support parameters in Expression.GetText(), in order to avoid doing text replace for ParsedRange.RangeText

aviv86 committed:
RavenDB-10538 : fix tests, add test case for facets with numbers
